xdsl/transforms/function_transformations.py

```diff
@@ -36,6 +40,70 @@ def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
             rewriter.has_done_action = True
 
 
+TIMER_START = "timer_start"
+TIMER_END = "timer_end"
+
+
+@dataclass
+class AddBenchTimersPattern(RewritePattern):
+    start_func_t: func.FunctionType
+    end_func_t: func.FunctionType
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
+        if (
+            not (top_level := op.parent_op())
+            or not isinstance(top_level, builtin.ModuleOp)
+            or top_level.parent
+        ):
+            return
```

Comment:
We currently don't have a use case for this.

Reply:
We're looking at annotating top-level functions that are the equivalent of a main function. Should prob add a check that it's not being called. Don't think the SymbolTable API supports this though by any chance?

Reply:
Would the name of the function to annotate be worth adding as a pass parameter? If it works with your flow it might be the easiest way to make it safe by default.

Reply:
Good suggestion, let me come back to that after the deadline if alright